Don't use RTLD_GLOBAL to load _C.

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#31162 Don't use RTLD_GLOBAL to load _C.**
* #31161 Uniformly apply Windows logic in cpp_extensions everywhere

This should help us resolve a multitude of weird segfaults and crashes
when PyTorch is imported along with other packages. Those would often
happen because libtorch symbols were exposed globally and could be used
as a source of relocations in shared libraries loaded after libtorch.

Based off of apaszke's original work at https://github.com/pytorch/pytorch/pull/28536

Fixes #3059.

Some of the subtleties in preparing this patch:

* Getting ASAN to play ball was a pain in the ass. The basic problem is that when we load with `RTLD_LOCAL`, we now may load a library multiple times into the address space; this happens when we have custom C++ extensions. Since the libraries are usually identical, this is usually benign, but it is technically undefined behavior and UBSAN hates it. I sprayed a few ways of getting things to "work" correctly: I preload libstdc++ (so that it is seen consistently over all library loads) and added turned off vptr checks entirely. Another possibility is we should have a mode where we use RTLD_GLOBAL to load _C, which would be acceptable in environments where you're sure C++ lines up correctly. There's a long comment in the test script going into more detail about this.
* Making some of our shared library dependencies load with `RTLD_LOCAL` breaks them. OpenMPI and MKL don't work; they play linker shenanigans to look up their symbols which doesn't work when loaded locally, and if we load a library with `RLTD_LOCAL` we aren't able to subsequently see it with `ctypes`. To solve this problem, we employ a clever device invented by apaszke: we create a dummy library `torch_global_deps` with dependencies on all of the libraries which need to be loaded globally, and then load that with `RTLD_GLOBAL`. As long as none of these libraries have C++ symbols, we can avoid confusion about C++ standard library.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D19262579](https://our.internmc.facebook.com/intern/diff/D19262579)

ezyang committed:
Don't use RTLD_GLOBAL to load _C.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

kostmo commented:
## CircleCI build failures summary
As of commit beb3b40:
* **1/1** failures introduced in this PR



## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=beb3b40e2adfe0086d8974cc4855336fe1d784ff).

### 1 new failure recognized by patterns
The following build failures don't appear to be due to upstream breakage:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/3915772) pytorch_linux_xenial_py3_6_gcc5_4_test (1/1)
**Step:** "Test" ([details](https://dr.pytorch.org/build-details.html?build_id=72795901))
```
Dec 12 03:30:09 Ran 31 tests in 21.967s
Dec 12 03:30:09 
Dec 12 03:30:09 FAILED (errors=1, skipped=7)
Dec 12 03:30:09 
Dec 12 03:30:09 Generating XML reports...
Dec 12 03:30:09 Traceback (most recent call last):
Dec 12 03:30:09   File "test/run_test.py", line 456, in <module>
Dec 12 03:30:09     main()
Dec 12 03:30:09   File "test/run_test.py", line 449, in main
Dec 12 03:30:09     raise RuntimeError(message)
Dec 12 03:30:09 RuntimeError: test_cpp_extensions failed!
Dec 12 03:30:09 + cleanup
Dec 12 03:30:09 + retcode=1
Dec 12 03:30:09 + set +x
Dec 12 03:30:09 =================== sccache compilation log ===================
Dec 12 03:30:09 ERROR:sccache::server: Compilation failed: Output { status: ExitStatus(ExitStatus(256)), stdout: "", stderr: "/tmp/torch_extensions/test_compilation_error_formatting/main.cpp: In function \'int main()\':\n/tmp/torch_extensions/test_compilation_error_formatting/main.cpp:2:23: error: expected \';\' before \'}\' token\n int main() { return 0 }\n                       ^\n" }
Dec 12 03:30:09 
Dec 12 03:30:09 =========== If your build fails, please take a look at the log above for possible reasons ===========
Dec 12 03:30:09 Compile requests                 43
Dec 12 03:30:09 Compile requests executed        25
Dec 12 03:30:09 Cache hits                       18
```


---
This comment was automatically generated by [Dr. CI](https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment).
Follow [this link to opt-out](https://dr.pytorch.org/admin/comments-opt-out.html) of these comments for your Pull Requests.


Please report bugs/suggestions on the [GitHub issue tracker](https://github.com/kostmo/circleci-failure-tracker/issues).

ezyang committed:
Update on "Don't use RTLD_GLOBAL to load _C."


Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Don't use RTLD_GLOBAL to load _C."


Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Don't use RTLD_GLOBAL to load _C."


Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Don't use RTLD_GLOBAL to load _C."


Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Don't use RTLD_GLOBAL to load _C."


Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Don't use RTLD_GLOBAL to load _C."


Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Don't use RTLD_GLOBAL to load _C."


Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Don't use RTLD_GLOBAL to load _C."


Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Don't use RTLD_GLOBAL to load _C."


Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Don't use RTLD_GLOBAL to load _C."


Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Don't use RTLD_GLOBAL to load _C."


Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Don't use RTLD_GLOBAL to load _C."


Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Don't use RTLD_GLOBAL to load _C."


Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

peterjc123 commented on torch/__init__.py:
```diff
@@ -33,61 +34,44 @@
+    if os.path.exists(NVTOOLSEXT_PATH):
+        nvtoolsext_lib_path = os.path.join(NVTOOLSEXT_PATH, 'bin', 'x64')
+    else:
+        nvtoolsext_lib_path = ''
```
A tip that maybe not directly related to this PR: We missed the path of CUDA here. To make it compatible with Python 3.8, we have to do that.

ezyang replied:
I'm happy to fix this, but I'm not exactly sure what the suggestion is here. What's the other CUDA path we missed?

peterjc123 replied:
Do we have a variable that records the version of the cuda used during build? If yes, then the answer is quite simple. For example, for CUDA 9.2, just add the following path.
```python
cuda_path = os.path.join(os.environ.get('CUDA_PATH_V9_2', r'C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v9.2'), 'bin')
```

ezyang committed:
Update on "Don't use RTLD_GLOBAL to load _C."


Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D19262579](https://our.internmc.facebook.com/intern/diff/D19262579)

[ghstack-poisoned]

ezyang committed:
Update on "Don't use RTLD_GLOBAL to load _C."


Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D19262579](https://our.internmc.facebook.com/intern/diff/D19262579)

[ghstack-poisoned]

ezyang committed:
Update on "Don't use RTLD_GLOBAL to load _C."


Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D19262579](https://our.internmc.facebook.com/intern/diff/D19262579)

[ghstack-poisoned]

ezyang committed:
Update on "Don't use RTLD_GLOBAL to load _C."


Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D19262579](https://our.internmc.facebook.com/intern/diff/D19262579)

[ghstack-poisoned]

ezyang committed:
Update on "Don't use RTLD_GLOBAL to load _C."


Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D19262579](https://our.internmc.facebook.com/intern/diff/D19262579)

[ghstack-poisoned]

ezyang committed:
Update on "Don't use RTLD_GLOBAL to load _C."


Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D19262579](https://our.internmc.facebook.com/intern/diff/D19262579)

[ghstack-poisoned]

albanD commented on .jenkins/pytorch/test.sh:
```diff
@@ -76,22 +76,51 @@ fi
+    # especially applies to type info, which is almost always weak.  This
+    # has implications for RTTI (which UBSAN is rightly flagging won't
+    # work), but in our codebase, we don't use RTTI (because it doesn't
+    # work in mobile).  However, UBSAN relies on UBSAN to detect vptr
```
nit: UBSAN relies on UBSAN ?

ezyang committed:
Update on "Don't use RTLD_GLOBAL to load _C."


This should help us resolve a multitude of weird segfaults and crashes
when PyTorch is imported along with other packages. Those would often
happen because libtorch symbols were exposed globally and could be used
as a source of relocations in shared libraries loaded after libtorch.

Fixes #3059.

Some of the subtleties in preparing this patch:

* Getting ASAN to play ball was a pain in the ass. The basic problem is that when we load with `RTLD_LOCAL`, we now may load a library multiple times into the address space; this happens when we have custom C++ extensions. Since the libraries are usually identical, this is usually benign, but it is technically undefined behavior and UBSAN hates it. I sprayed a few ways of getting things to "work" correctly: I preload libstdc++ (so that it is seen consistently over all library loads) and added turned off vptr checks entirely. Another possibility is we should have a mode where we use RTLD_GLOBAL to load _C, which would be acceptable in environments where you're sure C++ lines up correctly. There's a long comment in the test script going into more detail about this.
* Making some of our shared library dependencies load with `RTLD_LOCAL` breaks them. OpenMPI and MKL don't work; they play linker shenanigans to look up their symbols which doesn't work when loaded locally, and if we load a library with `RLTD_LOCAL` we aren't able to subsequently see it with `ctypes`. To solve this problem, we employ a clever device invented by apaszke: we create a dummy library `torch_global_deps` with dependencies on all of the libraries which need to be loaded globally, and then load that with `RTLD_GLOBAL`. As long as none of these libraries have C++ symbols, we can avoid confusion about C++ standard library.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D19262579](https://our.internmc.facebook.com/intern/diff/D19262579)

[ghstack-poisoned]

ezyang committed:
Update on "Don't use RTLD_GLOBAL to load _C."


This should help us resolve a multitude of weird segfaults and crashes
when PyTorch is imported along with other packages. Those would often
happen because libtorch symbols were exposed globally and could be used
as a source of relocations in shared libraries loaded after libtorch.

Fixes #3059.

Some of the subtleties in preparing this patch:

* Getting ASAN to play ball was a pain in the ass. The basic problem is that when we load with `RTLD_LOCAL`, we now may load a library multiple times into the address space; this happens when we have custom C++ extensions. Since the libraries are usually identical, this is usually benign, but it is technically undefined behavior and UBSAN hates it. I sprayed a few ways of getting things to "work" correctly: I preload libstdc++ (so that it is seen consistently over all library loads) and added turned off vptr checks entirely. Another possibility is we should have a mode where we use RTLD_GLOBAL to load _C, which would be acceptable in environments where you're sure C++ lines up correctly. There's a long comment in the test script going into more detail about this.
* Making some of our shared library dependencies load with `RTLD_LOCAL` breaks them. OpenMPI and MKL don't work; they play linker shenanigans to look up their symbols which doesn't work when loaded locally, and if we load a library with `RLTD_LOCAL` we aren't able to subsequently see it with `ctypes`. To solve this problem, we employ a clever device invented by apaszke: we create a dummy library `torch_global_deps` with dependencies on all of the libraries which need to be loaded globally, and then load that with `RTLD_GLOBAL`. As long as none of these libraries have C++ symbols, we can avoid confusion about C++ standard library.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D19262579](https://our.internmc.facebook.com/intern/diff/D19262579)

[ghstack-poisoned]

ezyang committed:
Update on "Don't use RTLD_GLOBAL to load _C."


This should help us resolve a multitude of weird segfaults and crashes
when PyTorch is imported along with other packages. Those would often
happen because libtorch symbols were exposed globally and could be used
as a source of relocations in shared libraries loaded after libtorch.

Fixes #3059.

Some of the subtleties in preparing this patch:

* Getting ASAN to play ball was a pain in the ass. The basic problem is that when we load with `RTLD_LOCAL`, we now may load a library multiple times into the address space; this happens when we have custom C++ extensions. Since the libraries are usually identical, this is usually benign, but it is technically undefined behavior and UBSAN hates it. I sprayed a few ways of getting things to "work" correctly: I preload libstdc++ (so that it is seen consistently over all library loads) and added turned off vptr checks entirely. Another possibility is we should have a mode where we use RTLD_GLOBAL to load _C, which would be acceptable in environments where you're sure C++ lines up correctly. There's a long comment in the test script going into more detail about this.
* Making some of our shared library dependencies load with `RTLD_LOCAL` breaks them. OpenMPI and MKL don't work; they play linker shenanigans to look up their symbols which doesn't work when loaded locally, and if we load a library with `RLTD_LOCAL` we aren't able to subsequently see it with `ctypes`. To solve this problem, we employ a clever device invented by apaszke: we create a dummy library `torch_global_deps` with dependencies on all of the libraries which need to be loaded globally, and then load that with `RTLD_GLOBAL`. As long as none of these libraries have C++ symbols, we can avoid confusion about C++ standard library.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D19262579](https://our.internmc.facebook.com/intern/diff/D19262579)

[ghstack-poisoned]

peterjc123 commented on torch/__init__.py:
```diff
@@ -60,18 +61,41 @@ def _load_global_deps():
+    # you load consistently use the same libstdc++, or you may have
+    # mysterious segfaults.
+    #
+    if platform.system() != 'Windows':
```
Maybe you need to move this if condition to L84, otherwise if `TORCH_USE_RTLD_GLOBAL` is accidentally defined in a Windows environment, it will not work.

peterjc123 commented on torch/__init__.py:
```diff
@@ -33,61 +35,81 @@
+
+# See Note [Global dependencies]
+def _load_global_deps():
+    if platform.system() == 'Windows':
```
What about defining a global variable `IS_WINDOWS` since it's used multiple times in this file?

ezyang replied:
I'll do this in a follow up

ezyang committed:
Update on "Don't use RTLD_GLOBAL to load _C."


This should help us resolve a multitude of weird segfaults and crashes
when PyTorch is imported along with other packages. Those would often
happen because libtorch symbols were exposed globally and could be used
as a source of relocations in shared libraries loaded after libtorch.

Fixes #3059.

Some of the subtleties in preparing this patch:

* Getting ASAN to play ball was a pain in the ass. The basic problem is that when we load with `RTLD_LOCAL`, we now may load a library multiple times into the address space; this happens when we have custom C++ extensions. Since the libraries are usually identical, this is usually benign, but it is technically undefined behavior and UBSAN hates it. I sprayed a few ways of getting things to "work" correctly: I preload libstdc++ (so that it is seen consistently over all library loads) and added turned off vptr checks entirely. Another possibility is we should have a mode where we use RTLD_GLOBAL to load _C, which would be acceptable in environments where you're sure C++ lines up correctly. There's a long comment in the test script going into more detail about this.
* Making some of our shared library dependencies load with `RTLD_LOCAL` breaks them. OpenMPI and MKL don't work; they play linker shenanigans to look up their symbols which doesn't work when loaded locally, and if we load a library with `RLTD_LOCAL` we aren't able to subsequently see it with `ctypes`. To solve this problem, we employ a clever device invented by apaszke: we create a dummy library `torch_global_deps` with dependencies on all of the libraries which need to be loaded globally, and then load that with `RTLD_GLOBAL`. As long as none of these libraries have C++ symbols, we can avoid confusion about C++ standard library.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D19262579](https://our.internmc.facebook.com/intern/diff/D19262579)

[ghstack-poisoned]

ezyang committed:
Update on "Don't use RTLD_GLOBAL to load _C."


This should help us resolve a multitude of weird segfaults and crashes
when PyTorch is imported along with other packages. Those would often
happen because libtorch symbols were exposed globally and could be used
as a source of relocations in shared libraries loaded after libtorch.

Fixes #3059.

Some of the subtleties in preparing this patch:

* Getting ASAN to play ball was a pain in the ass. The basic problem is that when we load with `RTLD_LOCAL`, we now may load a library multiple times into the address space; this happens when we have custom C++ extensions. Since the libraries are usually identical, this is usually benign, but it is technically undefined behavior and UBSAN hates it. I sprayed a few ways of getting things to "work" correctly: I preload libstdc++ (so that it is seen consistently over all library loads) and added turned off vptr checks entirely. Another possibility is we should have a mode where we use RTLD_GLOBAL to load _C, which would be acceptable in environments where you're sure C++ lines up correctly. There's a long comment in the test script going into more detail about this.
* Making some of our shared library dependencies load with `RTLD_LOCAL` breaks them. OpenMPI and MKL don't work; they play linker shenanigans to look up their symbols which doesn't work when loaded locally, and if we load a library with `RLTD_LOCAL` we aren't able to subsequently see it with `ctypes`. To solve this problem, we employ a clever device invented by apaszke: we create a dummy library `torch_global_deps` with dependencies on all of the libraries which need to be loaded globally, and then load that with `RTLD_GLOBAL`. As long as none of these libraries have C++ symbols, we can avoid confusion about C++ standard library.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D19262579](https://our.internmc.facebook.com/intern/diff/D19262579)

[ghstack-poisoned]

ezyang committed:
Update on "Don't use RTLD_GLOBAL to load _C."


This should help us resolve a multitude of weird segfaults and crashes
when PyTorch is imported along with other packages. Those would often
happen because libtorch symbols were exposed globally and could be used
as a source of relocations in shared libraries loaded after libtorch.

Fixes #3059.

Some of the subtleties in preparing this patch:

* Getting ASAN to play ball was a pain in the ass. The basic problem is that when we load with `RTLD_LOCAL`, we now may load a library multiple times into the address space; this happens when we have custom C++ extensions. Since the libraries are usually identical, this is usually benign, but it is technically undefined behavior and UBSAN hates it. I sprayed a few ways of getting things to "work" correctly: I preload libstdc++ (so that it is seen consistently over all library loads) and added turned off vptr checks entirely. Another possibility is we should have a mode where we use RTLD_GLOBAL to load _C, which would be acceptable in environments where you're sure C++ lines up correctly. There's a long comment in the test script going into more detail about this.
* Making some of our shared library dependencies load with `RTLD_LOCAL` breaks them. OpenMPI and MKL don't work; they play linker shenanigans to look up their symbols which doesn't work when loaded locally, and if we load a library with `RLTD_LOCAL` we aren't able to subsequently see it with `ctypes`. To solve this problem, we employ a clever device invented by apaszke: we create a dummy library `torch_global_deps` with dependencies on all of the libraries which need to be loaded globally, and then load that with `RTLD_GLOBAL`. As long as none of these libraries have C++ symbols, we can avoid confusion about C++ standard library.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D19262579](https://our.internmc.facebook.com/intern/diff/D19262579)

[ghstack-poisoned]

apaszke commented on torch/__init__.py:
```diff
@@ -33,61 +35,81 @@
+    here = os.path.abspath(__file__)
+    lib_path = os.path.join(os.path.dirname(here), 'lib', lib_name)
+
+    ctypes.CDLL(lib_path, mode=ctypes.RTLD_GLOBAL)
```
I've been going through this out of curiosity and it got me wondering if this doesn't lead to an eventual `dlclose`? Don't we have to stash this library handle somewhere?

ezyang replied:
https://stackoverflow.com/questions/359498/how-can-i-unload-a-dll-using-ctypes-in-python suggests that it doesn't

apaszke commented on torch/_utils_internal.py:
```diff
@@ -54,3 +54,4 @@ def get_source_lines_and_file(obj):
 
 TEST_MASTER_ADDR = '127.0.0.1'
 TEST_MASTER_PORT = 29500
+USE_RTLD_GLOBAL_WITH_LIBTORCH = False
```
Why do we have this constant if it's always false? Is this so that you can patch it in fbcode?

ezyang replied:
Yep, fbcode shenanigans
